src/storage/src/hummock/sstable/xor_filter.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use bytes::{Buf, BufMut};
+use itertools::Itertools;
+use xorf::{Filter, Xor16};
+
+use super::{FilterBuilder, Sstable};
+
+pub struct XorFilterBuilder {
+    key_hash_entries: Vec<u64>,
+}
+
+impl XorFilterBuilder {
+    pub fn new(capacity: usize) -> Self {
+        let key_hash_entries = if capacity > 0 {
+            Vec::with_capacity(capacity)
+        } else {
+            vec![]
+        };
+        Self { key_hash_entries }
+    }
+}
+
+impl FilterBuilder for XorFilterBuilder {
+    fn add_key(&mut self, key: &[u8], table_id: u32) {
+        self.key_hash_entries
+            .push(Sstable::hash_for_bloom_filter(key, table_id));
+    }
+
+    fn approximate_len(&self) -> usize {
+        self.key_hash_entries.len() * 5
```

Comment:
Why using `self.key_hash_entries.len() * 5`? I thought bpe for `xor16` is around 20bits (~3 Bytes)

Reply:
> Why using `self.key_hash_entries.len() * 5`? I thought bpe for `xor16` is around 20bits (~3 Bytes)

good catch. This value for bloom filter was `*4` in previous code so I will replace this `*5` with `*4`.